### PR TITLE
[Data views]: Add quick actions

### LINKED ISSUE #55096: Table: Quick-actions
This ticket is a stub to include designs for quick-actions in table views. 

"Quick-actions" include things like moving pages to the trash, or reseting a template. Additionally, they can potentially encapsulate much of the "quick edit" affordance on wp-admin tables, IE enabling users to edit post meta without engaging the full-screen editor.

### PR DESCRIPTION
**Pull request by ntsekouras**
## What?
Resolves: https://github.com/WordPress/gutenberg/issues/55096

This PR adds quick actions in data views items. The idea is that an item can have primary and secondary actions. The API was updated with a new `isEligible` property, which handles the rendering or not of an action per item. This is needed because not all of the items can have the same actions.

### Design notes
This PR will need some design tweaks - at least:
1. Handle how we want to show the actions if they have only primary or secondary actions. Do we want to reserve a space even if the item has no secondary actions? 🤔 
2. Check design in `grid` view

## Testing Instructions
1. Go to pages list in site editor and look around
4. Everything should be functional

## Screenshots or screencast 
Screenshot 2023-10-25 at 12 17 44 PM

### CODE REVIEW

File: packages/edit-site/src/components/actions/trash-post.js
Comment (oandregal) [inline packages/edit-site/src/components/actions/trash-post.js]: This made me think of this interaction in wp-admin:

https://github.com/WordPress/gutenberg/assets/583546/7ee9a7a3-ea07-4fa6-a30f-d14815ea7485

Perhaps we should add `deletePermanently` action and make it available for every page (just as secondary)?

`deletePermanently` would probably require a different interaction (modal to confirm). It could be valuable to showcase what actions can do.

File: packages/edit-site/src/components/actions/view-actions.js
Comment (oandregal) [inline packages/edit-site/src/components/actions/view-actions.js]: Nit pick. My brain favors homogeneity, so I think it'd be happier if we did either 1) define all actions in one file or 2) define one action per file.

File: packages/edit-site/src/components/actions/view-actions.js
Comment (oandregal) [inline packages/edit-site/src/components/actions/view-actions.js]: 👍 on `isEligible`

File: packages/edit-site/src/components/actions/view-actions.js
Comment (oandregal) [inline packages/edit-site/src/components/actions/view-actions.js]: What are your thoughts on `isQuickAction: boolean` instead of `type`? Or other way to put it: do you anticipate having tertiary/quaternary/etc. actions?

It sounds like `type` should be fine – I'm mostly curious.

File: packages/edit-site/src/components/actions/view-actions.js
Comment (oandregal) [inline packages/edit-site/src/components/actions/view-actions.js]: I checked how this is currently managed in wp-admin and only `trash` posts don't have a preview (private or scheduled ones do, for example). It sounds good as it is.

File: packages/edit-site/src/components/dataviews/view-list.js
Comment (oandregal) [inline packages/edit-site/src/components/dataviews/view-list.js:37]: Ah, nice one: these are not _field_ actions but _record_ actions. I guess _item_ actions works as fine, and we are already using item in `getValue` and `render` as well. 👍

Comment (oandregal) [reply packages/edit-site/src/components/dataviews/view-list.js:37]: Actually, I wouldn't mind if we called this component just `<Actions />` and `item-actions.js` was moved to `actions/index.js`. Though, this is a private component not meant to be public, so I don't have a strong opinion.

File: packages/edit-site/src/components/dataviews/item-actions.js
Comment (oandregal) [inline packages/edit-site/src/components/dataviews/item-actions.js]:
```suggestion
									isDestructive={ action.isDestructive }
```

I haven't seen any action implementing this, so I presume it hasn't been used yet?

File: packages/edit-site/src/components/dataviews/item-actions.js
Comment (oandregal) [inline packages/edit-site/src/components/dataviews/item-actions.js]:
```suggestion
						isDestructive={ action.isDestructive }
```

File: packages/edit-site/src/components/dataviews/item-actions.js
Comment (oandregal) [inline packages/edit-site/src/components/dataviews/item-actions.js:17]: Personal preference: I'd favor using `actions.forEach` for this kind of operation.

**Review by oandregal — COMMENTED** [review]
I presume you are waiting for design feedback to land this, I'll let others do that. From a code point of view, this could be merged – though I left some comments.

If you'd rather land parts of this PR before the design conversation settles, one thing we could do is merge everything but the `type` feature (more actions available, isEligible, renaming things).

File: packages/edit-site/src/components/dataviews/view-list.js
Comment (ntsekouras) [reply packages/edit-site/src/components/dataviews/view-list.js:37]: We also have the `view-actions` so we could leave as is for now. I think the folder Riad had created was about maybe reusing these actions elsewhere? 🤔 We can reconsider in the future.

File: packages/edit-site/src/components/actions/view-actions.js
Comment (ntsekouras) [reply packages/edit-site/src/components/actions/view-actions.js]: Boolean sounds good for now. If we need more types in the future we can always change again. I'll update tomorrow.

File: packages/edit-site/src/components/dataviews/item-actions.js
Comment (ntsekouras) [reply packages/edit-site/src/components/dataviews/item-actions.js]: Good catch here! I'll update manually because there is one more occurrence in code(not in this change set).

File: packages/edit-site/src/components/dataviews/item-actions.js
Comment (ntsekouras) [reply packages/edit-site/src/components/dataviews/item-actions.js]: The trash post action does.

File: packages/edit-site/src/components/actions/trash-post.js
Comment (ntsekouras) [reply packages/edit-site/src/components/actions/trash-post.js]: Exactly! I didn't add any of those for now, because we can start adding actions incrementally, after we land the main PR. What you mentioned though and the `restore` should be a quick follow up.

File: packages/edit-site/src/components/actions/view-actions.js
Comment (ntsekouras) [reply packages/edit-site/src/components/actions/view-actions.js]: I can do that. I'll update and move them to single file. If we need to, we can reconsider/split in the future.

**Comment by jameskoster** [comment]
> Handle how we want to show the actions if they have only primary or secondary actions

Good question, and I'd welcome feedback on it. That said we may not need to prioritise it here since all pages should have the 'Edit' and 'Rename' actions. 

Some minor notes:

* Trash icon doesn't need to be red.
* Would it be trivial to add an 'Undo' affordance to the trash confirmation snackbar?
* Buttons should be the 32px size variant.
* Should the "View" action open in a new tab as the icon suggests? It doesn't in wp-admin, so perhaps we should be consistent with that. If so we may need to consider a different icon.

**Comment by ntsekouras** [comment]
@jameskoster I addressed the feedback, but didn't handle this one:
>Would it be trivial to add an 'Undo' affordance to the trash confirmation snackbar?

Let's do this in a follow up, to get the actions API in. This way we can iterate and in parallel others could add more actions.

>Should the "View" action open in a new tab as the icon suggests?

The page opens in the same page for consistency with the current pages list. If you have an icon to suggest, we can update easily.

**Review by youknowriad — APPROVED** [review]
LGTM 👍

**Pull request merged**
